subsys/usb/host/class/usbh_uvc.c

```diff
@@ -104,115 +322,325 @@ const void *usbh_uvc_desc_get_vc_end(const struct usb_if_descriptor *if_desc,
+	for (; desc != NULL && (uint8_t *)desc < (uint8_t *)desc_end;
+	     desc = usbh_desc_get_next(desc, desc_end)) {
 		if (desc->bDescriptorType == USB_DESC_INTERFACE ||
-		    desc->bDescriptorType == USB_DESC_INTERFACE_ASSOC ||
-		    desc->bDescriptorType == 0) {
+		    desc->bDescriptorType == USB_DESC_INTERFACE_ASSOC) {
 			break;
-		} else if (desc->bDescriptorType == USB_DESC_CS_INTERFACE) {
-			const struct uvc_if_descriptor *const if_desc = (const void *)desc;
+		}
 
-			switch (if_desc->bDescriptorSubtype) {
-			case UVC_VS_INPUT_HEADER:
-				LOG_DBG("VideoStreaming interface: Input header");
-				break;
-			case UVC_VS_OUTPUT_HEADER:
-				LOG_DBG("VideoStreaming interface: Output header");
-				break;
-			case UVC_VS_FORMAT_UNCOMPRESSED:
-				LOG_DBG("VideoStreaming interface: Uncompressed format");
-				break;
-			case UVC_VS_FORMAT_MJPEG:
-				LOG_DBG("VideoStreaming interface: MJPEG format");
-				break;
-			case UVC_VS_FRAME_UNCOMPRESSED:
-				LOG_DBG("VideoStreaming interface: Uncompressed Frame");
-				break;
-			case UVC_VS_FRAME_MJPEG:
-				LOG_DBG("VideoStreaming interface: MJPEG Frame");
-				break;
-			case UVC_VS_COLORFORMAT:
-				LOG_DBG("VideoStreaming interface: Color");
-				break;
-			default:
-				LOG_DBG("VideoStreaming descriptor: unknown subtype %u, skipping",
-					if_desc->bDescriptorSubtype);
-			}
-		} else if (desc->bDescriptorType == USB_DESC_ENDPOINT) {
+		if (desc->bDescriptorType == USB_DESC_ENDPOINT) {
 			const struct usb_ep_descriptor *const ep_desc = (const void *)desc;
+			LOG_DBG("VideoStreaming Endpoint: 0x%02x", ep_desc->bEndpointAddress);
+			continue;
+		}
 
-			LOG_DBG("VideoStreaming Endpoint", ep_desc->bEndpointAddress);
-		} else {
+		if (desc->bDescriptorType != USB_DESC_CS_INTERFACE) {
 			LOG_WRN("VideoStreaming descriptor: unknown type %u, skipping",
 				desc->bDescriptorType);
+			continue;
+		}
+
+		const struct uvc_cs_descriptor_header *const cs_desc = (const void *)desc;
+
+		switch (cs_desc->bDescriptorSubtype) {
+		case UVC_VS_INPUT_HEADER: {
+			struct uvc_stream_header_descriptor *header_desc =
+				(struct uvc_stream_header_descriptor *)desc;
+
+			if (desc->bLength < sizeof(struct uvc_stream_header_descriptor)) {
+				LOG_ERR("Invalid VS input header descriptor length: %u",
+					desc->bLength);
+				return -EINVAL;
+			}
+
+			uvc_dev->uvc_descriptors.vs_input_header = header_desc;
+			LOG_DBG("Found VS input header: formats=%u", header_desc->bNumFormats);
+			break;
+		}
+
+		case UVC_VS_OUTPUT_HEADER: {
+			struct uvc_stream_header_descriptor *header_desc =
+				(struct uvc_stream_header_descriptor *)desc;
+
+			if (desc->bLength < sizeof(struct uvc_stream_header_descriptor)) {
+				LOG_ERR("Invalid VS output header descriptor length: %u",
+					desc->bLength);
+				return -EINVAL;
+			}
+
+			uvc_dev->uvc_descriptors.vs_output_header = header_desc;
+			LOG_DBG("Found VS output header: formats=%u", header_desc->bNumFormats);
+			break;
+		}
```

Comment:
Stream output is for when video stream goes host -> device, for instance a video projector.
It is rarely used and maybe it is possible to ignore it for first implementation to reduce the code?